nixos/modules/services/web-apps/nextcloud-backup.nix

```diff
@@ -0,0 +1,199 @@
+      exit 1
+    fi
+    if [ "$(id -u)" != "0" ] ;then
+      echo "This script has to be run as root, aborting."
+      exit 1
+    fi
+    # Read backup metadata
+    NEXTCLOUD_MAJOR_VERSION=$(grep "^Nextcloud Version:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Nextcloud Version: ([0-9]+)\..*/\1/")
+    NEXTCLOUD_DB_TYPE=$(grep "^Database Type:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Database Type: //")
+    BACKUP_FORMAT_VERSION=$(grep "^Backup Format Version:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Backup Format Version: //")
+    # A few sanity checks to make sure that it's possible to restore the backup.
+    if [ ! "$BACKUP_FORMAT_VERSION" = "1" ] ;then
+      # In case there are backwards-incompatible changes to this script in the future
+      echo "Unknown backup format version $BACKUP_FORMAT_VERSION, aborting."
+      exit 1
+    fi
+    if [ ! "$NEXTCLOUD_MAJOR_VERSION" = "${lib.versions.major cfg.package.version}" ] ;then
+      echo "The backup is for Nextcloud $NEXTCLOUD_MAJOR_VERSION but Nextcloud ${lib.versions.major cfg.package.version} is installed."
+      echo "Please change to Nextcloud $NEXTCLOUD_MAJOR_VERSION before restoring (and then possibly upgrade), aborting."
+      exit 1
+    fi
+    if [ ! "$NEXTCLOUD_DB_TYPE" = "${cfg.config.dbtype}" ] ;then
+      echo "The backup is for the database $NEXTCLOUD_DB_TYPE but the current installation uses ${cfg.config.dbtype}, aborting."
+      exit 1
+    fi
+    ${checkBackup}
+    if [ ! -e "${cfg.datadir}/config/override.config.php" ] ;then
+      echo "Expected to find the file ${cfg.datadir}/config/override.config.php but it seems to be missing."
+      echo "It appears that nextcloud has not been set up correctly, aborting."
+      exit 1
+    fi
+    printf "WARNING: Restoring the backup will overwrite the current nextcloud instance. Continue? (y/n):"
+    read answer
+    if [ "$answer" = "''${answer#[Yy]}" ] ;then
+        echo "Restore cancelled, aborting."
+        exit 1
+    fi
+    echo "Enabling maintenance mode..."
+    ${cfg.occ}/bin/nextcloud-occ maintenance:mode --on
+    echo "Restoring config and manually installed apps..."
+    ${restoreRsyncCommand} $1/nextcloud-backup/config.php ${cfg.datadir}/config/
+    ${restoreRsyncCommand} $1/nextcloud-backup/store-apps ${cfg.home}
+    echo "Restoring data directory..."
+    ${restoreRsyncCommand} $1/nextcloud-backup/data ${cfg.datadir}
+    echo "Restoring database..."
+    ${dbBackupRestore.restoreCommand}
+    echo "Disabling maintenance mode..."
+    ${cfg.occ}/bin/nextcloud-occ maintenance:mode --off
+    echo "Nextcloud backup successfully restored."
+    echo "If the backup was not very recent, it might make sense to run \"nextcloud-occ maintenance:data-fingerprint\"."
+    echo "See https://docs.nextcloud.com/server/latest/admin_manual/maintenance/restore.html#synchronising-with-clients-after-data-recovery for more information."
+  '';
+in
+{
+  options.services.nextcloud = {
```

Comment:
I must say, I'm not comfortable with having a non-official backup solution in the `services.nextcloud` namespace.

Would it be OK for everyone to move this into `services.nextcloudBackup` or something like this?
I think `mysql`/`postgresql` are good precedents for that.